### PR TITLE
Add support of run time flags for enable Castanets

### PR DESCRIPTION
**Pull request by santhoshkerur**
This change adds runtime flag to enable castanets.
Runtime flags --enable-forking, --server-address has been removed.
Castanets can be verified as below,

Standalone mode:
chrome <url>

Castanets mode:
Browser: chrome <url> --enable-castanets
Renderer: chrome --type=renderer --enable-castanets=<IP Address>

Signed-off-by: Santhoshkumar Kerur <santhosh.vk@samsung.com>
Signed-off-by: suyambu.rm <suyambu.rm@samsung.com>

### CODE REVIEW

File: components/discardable_memory/client/client_discardable_shared_memory_manager.cc [hunk 1]
```diff
@@ -360,33 +362,37 @@ ClientDiscardableSharedMemoryManager::AllocateLockedDiscardableSharedMemory(
...
+  if (base::Castanets::IsEnabled()) {
+    std::unique_ptr<base::DiscardableSharedMemory> memory(
+        new base::DiscardableSharedMemory);
+    if (!memory->CreateAndMap(size)) {
+      region = base::UnsafeSharedMemoryRegion();
+      return nullptr;
+    }
+    //*shared_memory_region = memory->DuplicateRegion();
+    // Close file descriptor to avoid running out.
+    // memory->Close(); needed ?
```
Comment (suyambulingamrm) [inline components/discardable_memory/client/client_discardable_shared_memory_manager.cc]: May be we can remove these commented lines too.

File: content/browser/renderer_host/render_process_host_impl.cc [hunk 2]
```diff
@@ -3105,7 +3109,8 @@ void RenderProcessHostImpl::PropagateBrowserCommandLineToRenderer(
     switches::kIpcFuzzerTestcase,
 #endif
 #if defined(CASTANETS)
-    switches::kServerAddress,
+    switches::kEnableCastanets,
+   //switches::kServerAddress,
```
Comment (suyambulingamrm) [inline content/browser/renderer_host/render_process_host_impl.cc]: can remove commented lines

**Review by suyambulingamrm — COMMENTED** [review]
There are many changes specific to ANDROID and CASTANETS in below files..
We need to cross check if runtime checks are needed for those too.

gpu/ipc/common/gpu_memory_buffer_support.cc
gpu/ipc/service/gpu_memory_buffer_factory.cc
ui/gfx/mojo/buffer_types_struct_traits.cc
content/browser/frame_host/frame_tree_node.cc
content/browser/frame_host/render_frame_host_impl.cc	
content/browser/android/content_startup_flags.cc
content/browser/child_process_launcher_helper_android.cc
content/public/browser/content_browser_client.cc
content/public/common/use_zoom_for_dsf_policy.cc
content/public/common/content_features.cc
content/renderer/render_widget.cc
content/renderer/input/render_widget_input_handler.cc
third_party/blink/renderer/core/loader/resource/script_resource.cc
third_party/blink/public/common/page/launching_process_state.h
mojo/public/cpp/system/platform_handle.cc
mojo/core/platform_handle_utils.cc
mojo/core/broker_castanets.cc
mojo/core/core.cc
mojo/core/broker_host.cc
mojo/core/shared_buffer_dispatcher.cc

File: components/discardable_memory/client/client_discardable_shared_memory_manager.cc (hunk 1 above)
Comment (santhoshkerur) [reply components/discardable_memory/client/client_discardable_shared_memory_manager.cc]: okay.

File: content/browser/renderer_host/render_process_host_impl.cc (hunk 2 above)
Comment (santhoshkerur) [reply content/browser/renderer_host/render_process_host_impl.cc]: okay. Done.

**Comment by santhoshkerur** [comment]
> There are many changes specific to ANDROID and CASTANETS in below files..
> We need to cross check if runtime checks are needed for those too.
> 
> gpu/ipc/common/gpu_memory_buffer_support.cc
> gpu/ipc/service/gpu_memory_buffer_factory.cc
> ui/gfx/mojo/buffer_types_struct_traits.cc
> content/browser/frame_host/frame_tree_node.cc
> content/browser/frame_host/render_frame_host_impl.cc
> content/browser/android/content_startup_flags.cc
> content/browser/child_process_launcher_helper_android.cc
> content/public/browser/content_browser_client.cc
> content/public/common/use_zoom_for_dsf_policy.cc
> content/public/common/content_features.cc
> content/renderer/render_widget.cc
> content/renderer/input/render_widget_input_handler.cc
> third_party/blink/renderer/core/loader/resource/script_resource.cc
> third_party/blink/public/common/page/launching_process_state.h
> mojo/public/cpp/system/platform_handle.cc
> mojo/core/platform_handle_utils.cc
> mojo/core/broker_castanets.cc
> mojo/core/core.cc
> mojo/core/broker_host.cc
> mojo/core/shared_buffer_dispatcher.cc

@suyambulingamrm , Files have changed. Thanks for review and Support as co-contributor.

**Comment by suyambulingamrm** [comment]
+1.
Verified on desktop, android.
Below are not part of this patch yet.
1. Minimise castanets main activity only when --enable-castanets is provided. (Inprogress will be part of new patchset)
2.  We were using BGRA as Skia format for castanets to make android renderer compatible with chrome/efl browser. Because of this we see incorrect colors on some pages, when we run android browser without runtime flag. (TODO)

File: components/services/font/public/cpp/font_service_thread.cc [hunk 3]
```diff
@@ -144,8 +144,25 @@ scoped_refptr<MappedFontFile> FontServiceThread::OpenStream(
...
+    task_runner_->PostTask(
+        FROM_HERE, base::BindOnce(&FontServiceThread::OpenStreamImpl, this,
+                                  &done_event, &stream_file, identity.fID));
+    done_event.Wait();
+
+    if (!stream_file.IsValid()) {
+      // The font-service may have been killed.
+      return nullptr;
+    }
+  }
```
Comment (Sunwoo-Nam) [inline components/services/font/public/cpp/font_service_thread.cc:167]: Is it possible to remove duplicate code?

File: cc/raster/staging_buffer_pool.cc [hunk 4]
```diff
@@ -261,7 +266,15 @@ std::unique_ptr<StagingBuffer> StagingBufferPool::AcquireStagingBuffer(
...
-    ri->Finish();
+    if (base::Castanets::IsEnabled()) {
+      ri->Finish();
+    } else {
+      // Early out if query isn't used, or if query isn't complete yet.  Query
+      // is created in OneCopyRasterBufferProvider::CopyOnWorkerThread().
+      if (!busy_buffers_.front()->query_id ||
+          !CheckForQueryResult(ri, busy_buffers_.front()->query_id))
+        break;
+    }
```
Comment (Sunwoo-Nam) [inline cc/raster/staging_buffer_pool.cc:279]: Is it possible to remove duplicate code?

File: content/renderer/render_widget.cc [hunk 5]
```diff
@@ -1664,7 +1668,11 @@ void RenderWidget::IntrinsicSizingInfoChanged(
 void RenderWidget::DidMeaningfulLayout(blink::WebMeaningfulLayout layout_type) {
   if (layout_type == blink::WebMeaningfulLayout::kVisuallyNonEmpty) {
 #if defined(CASTANETS)
-    Send(new WidgetHostMsg_DidFirstVisuallyNonEmptyPaint(routing_id_));
+    if (base::Castanets::IsEnabled())
+      Send(new WidgetHostMsg_DidFirstVisuallyNonEmptyPaint(routing_id_));
+    else
+      QueueMessage(
+          new WidgetHostMsg_DidFirstVisuallyNonEmptyPaint(routing_id_));
```
Comment (Sunwoo-Nam) [inline content/renderer/render_widget.cc:1676]: Is it possible to remove duplicate code?

File: gpu/command_buffer/client/implementation_base.cc [hunk 6]
```diff
@@ -261,16 +265,20 @@ gpu::ContextResult ImplementationBase::Initialize(
...
+    // Call this api to synchronize the result shared memory.
+    helper_->SyncResultData(GetResultShmId(), GetResultShmOffset(),
+                            kMaxSizeOfSimpleResult);
+    helper_->Finish();
+
+    // Wait for command execution result.
+    mojo::WaitSyncSharedMemory(transfer_buffer_->shared_memory_guid());
+  } else {
+    helper_->Finish();
+  }
```
Comment (Sunwoo-Nam) [inline gpu/command_buffer/client/implementation_base.cc]: Is it possible to remove duplicate code?

File: mojo/core/broker_host.cc [hunk 7]
```diff
@@ -74,10 +78,15 @@ bool BrokerHost::SendChannel(PlatformHandle handle) {
...
-  data->port = -1;
+  Channel::MessagePtr message;
+  if (base::Castanets::IsEnabled()) {
+    InitData* data;
+    Channel::MessagePtr message =
+        CreateBrokerMessage(BrokerMessageType::INIT, 1, 0, &data);
+    data->port = -1;
+  } else {
+    message = CreateBrokerMessage(BrokerMessageType::INIT, 1, nullptr);
+  }
```
Comment (Sunwoo-Nam) [inline mojo/core/broker_host.cc]: Is it possible to remove duplicate code?

File: mojo/core/channel_posix.cc [hunk 8]
```diff
@@ -504,13 +506,18 @@ class ChannelPosix : public Channel,
...
+                                  &socket_);
+        if (secure_connection_) {
+          ssl_.reset(AcceptSSLConnection(socket_.get()));
+          CHECK(ssl_);
+          LOG(INFO) << "SSL Server Connection Success - socket: "
+                    << socket_.get();
+        }
+      } else {
+        AcceptSocketConnection(server_.platform_handle().GetFD().get(),
+                               &socket_);
```
Comment (Sunwoo-Nam) [inline mojo/core/channel_posix.cc:521]: Is it possible to remove duplicate code?

File: mojo/core/node_controller.cc [hunk 9]
```diff
@@ -250,17 +253,32 @@ void NodeController::AcceptBrokerClientInvitation(
...
+    PlatformChannelEndpoint endpoint = broker_->GetInviterEndpoint();
+    if (!endpoint.is_valid()) {
+      // Most likely the inviter's side of the channel has already been closed
+      // and the broker was unable to negotiate a NodeChannel pipe. In this case
+      // we can cancel our connection to our inviter.
+      DVLOG(1) << "Cannot connect to invalid inviter channel.";
+      CancelPendingPortMerges();
+      return;
+    }
+    connection_params = ConnectionParams(std::move(endpoint));
```
Comment (Sunwoo-Nam) [inline mojo/core/node_controller.cc:283]: Is it possible to remove duplicate code?

File: mojo/core/node_controller.cc [hunk 10]
```diff
@@ -487,17 +516,31 @@ void NodeController::SendBrokerClientInvitationOnIOThread(
...
+  } else {
     PlatformChannel node_channel;
     node_connection_params = ConnectionParams(node_channel.TakeLocalEndpoint());
+    // BrokerHost owns itself.
+    broker_host =
+        new BrokerHost(target_process.get(), std::move(connection_params),
+                       process_error_callback);
     channel_ok = broker_host->SendChannel(
         node_channel.TakeRemoteEndpoint().TakePlatformHandle());
   }
```
Comment (Sunwoo-Nam) [inline mojo/core/node_controller.cc:535]: Is it possible to remove duplicate code?

File: mojo/core/node_controller.cc [hunk 11]
```diff
@@ -1299,36 +1350,42 @@ void NodeController::OnIntroduce(const ports::NodeName& from_node,
...
-                              Channel::HandlePolicy::kAcceptHandles,
-                              io_task_runner_, ProcessErrorCallback());
-      DVLOG(1) << "Adding new peer " << name << " via broker introduction.";
-      AddPeer(name, channel, true /* start_channel */);
+    } else {
+      node_->LostConnectionToNode(name);
+      DVLOG(1) << "Could not be introduced to peer " << name;
+      base::AutoLock lock(peers_lock_);
+      pending_peer_messages_.erase(name);
     }
```
Comment (Sunwoo-Nam) [inline mojo/core/node_controller.cc:1371]: Is it possible to remove duplicate code?

File: mojo/public/cpp/system/invitation.cc [hunk 12]
```diff
@@ -97,17 +101,24 @@ void SendInvitation(ScopedInvitationHandle invitation,
...
+    options.secure_connection = secure_connection;
+    endpoint.secure_connection = secure_connection;
+    result = MojoSendInvitation(invitation.get().value(), &process_handle,
+                                &endpoint, error_handler, error_handler_context,
+                                &options, tcp_success_callback);
+  } else {
+    result =
+        MojoSendInvitation(invitation.get().value(), &process_handle, &endpoint,
+                           error_handler, error_handler_context, &options);
+  }
```
Comment (Sunwoo-Nam) [inline mojo/public/cpp/system/invitation.cc:121]: Is it possible to remove duplicate code?

**Review by Sunwoo-Nam — COMMENTED** [review]
I commented about duplicate codes, but it is not critical.

**Comment by suyambulingamrm** [comment]
> I commented about duplicate codes, but it is not critical.

Done.

**Review by Sunwoo-Nam — APPROVED** [review]
Looks good to me.

**Pull request merged**